--- a/net/minecraft/launcher/ui/popups/login/LogInPopup.java
+++ b/net/minecraft/launcher/ui/popups/login/LogInPopup.java
@@ -11,6 +11,7 @@
 import java.awt.image.BufferedImage;
 import java.io.IOException;
 import java.io.InputStream;
+import java.lang.reflect.InvocationTargetException;
 import javax.imageio.ImageIO;
 import javax.swing.Box;
 import javax.swing.BoxLayout;
@@ -21,11 +22,12 @@
 import javax.swing.JProgressBar;
 import javax.swing.SwingUtilities;
 import javax.swing.border.EmptyBorder;
+
+import me.nepnep.msa4legacy.MicrosoftAuth;
+import me.nepnep.msa4legacy.HasLoginPopup;
+import me.nepnep.msa4legacy.MicrosoftAccount;
 import net.minecraft.launcher.Launcher;
 import net.minecraft.launcher.LauncherConstants;
-import net.minecraft.launcher.ui.popups.login.AuthErrorForm;
-import net.minecraft.launcher.ui.popups.login.ExistingUserListForm;
-import net.minecraft.launcher.ui.popups.login.LogInForm;
 
 public class LogInPopup
 extends JPanel
@@ -38,17 +40,36 @@
     private final JButton loginButton = new JButton("Log In");
     private final JButton registerButton = new JButton("Register");
     private final JProgressBar progressBar = new JProgressBar();
+    public final JButton msaButton = new JButton("Microsoft Account");
+    public final HasLoginPopup msaLogInForm;
+    public final JPanel buttonPanel = new JPanel();
 
-    public LogInPopup(Launcher minecraftLauncher, Callback callback) {
+    public LogInPopup(Launcher minecraftLauncher, Callback callback, MicrosoftAuth msa, ClassLoader cl) {
         super(true);
         this.minecraftLauncher = minecraftLauncher;
         this.callback = callback;
         this.errorForm = new AuthErrorForm(this);
-        this.existingUserListForm = new ExistingUserListForm(this);
+        this.existingUserListForm = new ExistingUserListForm(this, msa);
+        try {
+            this.msaLogInForm = (HasLoginPopup) cl.loadClass("me.nepnep.msa4legacy.patches.MSALogInForm")
+                    .getConstructor(this.getClass(), MicrosoftAuth.class)
+                    .newInstance(this, msa);
+        } catch (InstantiationException e) {
+            throw new RuntimeException(e);
+        } catch (IllegalAccessException e) {
+            throw new RuntimeException(e);
+        } catch (InvocationTargetException e) {
+            throw new RuntimeException(e);
+        } catch (NoSuchMethodException e) {
+            throw new RuntimeException(e);
+        } catch (ClassNotFoundException e) {
+            throw new RuntimeException(e);
+        }
         this.logInForm = new LogInForm(this);
         this.createInterface();
         this.loginButton.addActionListener(this);
         this.registerButton.addActionListener(this);
+        this.msaButton.addActionListener(this);
     }
 
     protected void createInterface() {
@@ -73,11 +94,13 @@
         }
         this.add(this.errorForm);
         this.add(this.logInForm);
+        this.add(this.msaLogInForm);
         this.add(Box.createVerticalStrut(15));
-        JPanel buttonPanel = new JPanel();
-        buttonPanel.setLayout(new GridLayout(1, 2, 10, 0));
+        buttonPanel.setLayout(new GridLayout(1, 3, 10, 0));
         buttonPanel.add(this.registerButton);
         buttonPanel.add(this.loginButton);
+        msaLogInForm.setVisible(false);
+        buttonPanel.add(msaButton);
         this.add(buttonPanel);
         this.progressBar.setIndeterminate(true);
         this.progressBar.setVisible(false);
@@ -90,6 +113,8 @@
             this.logInForm.tryLogIn();
         } else if (e.getSource() == this.registerButton) {
             OperatingSystem.openLink(LauncherConstants.URL_REGISTER);
+        } else if (e.getSource() == msaButton) {
+            msaLogInForm.setVisible(true);
         }
     }
 
@@ -128,7 +153,11 @@
     }
 
     public void setLoggedIn(String uuid) {
-        this.callback.onLogIn(uuid);
+        this.callback.onLogIn(uuid, null);
+    }
+    
+    public void setLoggedIn(MicrosoftAccount msa) {
+        callback.onLogIn(null, msa);
     }
 
     public void repack() {
@@ -139,7 +168,6 @@
     }
 
     public static interface Callback {
-        public void onLogIn(String var1);
+        public void onLogIn(String var1, MicrosoftAccount msa);
     }
 }
-
