--- a/net/minecraft/launcher/game/GameLaunchDispatcher.java
+++ b/net/minecraft/launcher/game/GameLaunchDispatcher.java
@@ -13,8 +13,10 @@
 import com.mojang.launcher.updater.VersionSyncInfo;
 import java.io.File;
 import java.util.concurrent.locks.ReentrantLock;
+
+import me.nepnep.msa4legacy.MicrosoftAccount;
+import me.nepnep.msa4legacy.MicrosoftAuth;
 import net.minecraft.launcher.Launcher;
-import net.minecraft.launcher.game.MinecraftGameRunner;
 import net.minecraft.launcher.profile.Profile;
 import net.minecraft.launcher.profile.ProfileManager;
 
@@ -24,11 +26,13 @@
     private final String[] additionalLaunchArgs;
     private final ReentrantLock lock = new ReentrantLock();
     private final BiMap<UserAuthentication, MinecraftGameRunner> instances = HashBiMap.create();
+    private final MicrosoftAuth msa;
     private boolean downloadInProgress = false;
 
-    public GameLaunchDispatcher(Launcher launcher, String[] additionalLaunchArgs) {
+    public GameLaunchDispatcher(Launcher launcher, String[] additionalLaunchArgs, MicrosoftAuth msa) {
         this.launcher = launcher;
         this.additionalLaunchArgs = additionalLaunchArgs;
+        this.msa = msa;
     }
 
     /*
@@ -38,8 +42,10 @@
         UserAuthentication user;
         ProfileManager profileManager = this.launcher.getProfileManager();
         Profile profile = profileManager.getProfiles().isEmpty() ? null : profileManager.getSelectedProfile();
-        UserAuthentication userAuthentication = user = profileManager.getSelectedUser() == null ? null : profileManager.getAuthDatabase().getByUUID(profileManager.getSelectedUser());
-        if (user == null || !user.isLoggedIn() || profile == null || this.launcher.getLauncher().getVersionManager().getVersions(profile.getVersionFilter()).isEmpty()) {
+        String selectedUser = profileManager.getSelectedUser();
+        UserAuthentication userAuthentication = user = selectedUser == null ? null : profileManager.getAuthDatabase().getByUUID(profileManager.getSelectedUser());
+        MicrosoftAccount msa = selectedUser == null ? null : profileManager.getAuthDatabase().getMSAByUUID(selectedUser);
+        if (((user == null || !user.isLoggedIn()) && msa == null) || profile == null || this.launcher.getLauncher().getVersionManager().getVersions(profile.getVersionFilter()).isEmpty()) {
             return PlayStatus.LOADING;
         }
         this.lock.lock();
@@ -56,10 +62,10 @@
         finally {
             this.lock.unlock();
         }
-        if (user.getSelectedProfile() == null) {
+        if (user != null && user.getSelectedProfile() == null) {
             return PlayStatus.CAN_PLAY_DEMO;
         }
-        if (user.canPlayOnline()) {
+        if ((user != null && user.canPlayOnline()) || msa != null) {
             return PlayStatus.CAN_PLAY_ONLINE;
         }
         return PlayStatus.CAN_PLAY_OFFLINE;
@@ -93,7 +99,7 @@
         final Profile profile = profileManager.getSelectedProfile();
         UserAuthentication user = profileManager.getSelectedUser() == null ? null : profileManager.getAuthDatabase().getByUUID(profileManager.getSelectedUser());
         final String lastVersionId = profile.getLastVersionId();
-        final MinecraftGameRunner gameRunner = new MinecraftGameRunner(this.launcher, this.additionalLaunchArgs);
+        final MinecraftGameRunner gameRunner = new MinecraftGameRunner(this.launcher, this.additionalLaunchArgs, this.msa);
         gameRunner.setStatus(GameInstanceStatus.PREPARING);
         this.lock.lock();
         try {
@@ -194,4 +200,3 @@
         }
     }
 }
-
