--- a/net/minecraft/launcher/Launcher.java
+++ b/net/minecraft/launcher/Launcher.java
@@ -18,6 +18,7 @@
 import com.mojang.launcher.versions.Version;
 import com.mojang.util.UUIDTypeAdapter;
 import java.io.File;
+import java.io.FileFilter;
 import java.io.IOException;
 import java.net.PasswordAuthentication;
 import java.net.Proxy;
@@ -37,9 +38,8 @@
 import joptsimple.OptionException;
 import joptsimple.OptionParser;
 import joptsimple.OptionSet;
-import net.minecraft.launcher.LauncherConstants;
-import net.minecraft.launcher.MinecraftUserInterface;
-import net.minecraft.launcher.SwingUserInterface;
+import me.nepnep.msa4legacy.MicrosoftAccount;
+import me.nepnep.msa4legacy.MicrosoftAuth;
 import net.minecraft.launcher.game.GameLaunchDispatcher;
 import net.minecraft.launcher.game.MinecraftReleaseType;
 import net.minecraft.launcher.game.MinecraftReleaseTypeFactory;
@@ -71,6 +71,7 @@
     private final ProfileManager profileManager;
     private final Gson gson = new Gson();
     private final GameLaunchDispatcher launchDispatcher;
+    private final MicrosoftAuth msAuth;
     private boolean winTenHack = false;
     private UUID clientToken = UUID.randomUUID();
     private String requestedUser;
@@ -79,11 +80,11 @@
         return INSTANCE;
     }
 
-    public Launcher(JFrame frame, File workingDirectory, Proxy proxy, PasswordAuthentication proxyAuth, String[] args) {
-        this(frame, workingDirectory, proxy, proxyAuth, args, 0);
+    public Launcher(JFrame frame, File workingDirectory, Proxy proxy, PasswordAuthentication proxyAuth, String[] args, ClassLoader cl) {
+        this(frame, workingDirectory, proxy, proxyAuth, args, 0, cl);
     }
 
-    public Launcher(JFrame frame, File workingDirectory, Proxy proxy, PasswordAuthentication proxyAuth, String[] args, Integer bootstrapVersion) {
+    public Launcher(JFrame frame, File workingDirectory, Proxy proxy, PasswordAuthentication proxyAuth, String[] args, Integer bootstrapVersion, ClassLoader cl) {
         INSTANCE = this;
         this.setupErrorHandling();
         this.bootstrapVersion = bootstrapVersion;
@@ -105,8 +106,18 @@
         LOGGER.info("System.getProperty('java.vendor') == '" + System.getProperty("java.vendor") + "'");
         LOGGER.info("System.getProperty('sun.arch.data.model') == '" + System.getProperty("sun.arch.data.model") + "'");
         LOGGER.info("proxy == " + proxy);
-        this.launchDispatcher = new GameLaunchDispatcher(this, this.processArgs(args));
         this.launcher = new com.mojang.launcher.Launcher(this.userInterface, workingDirectory, proxy, proxyAuth, new MinecraftVersionManager(new LocalVersionList(workingDirectory), new RemoteVersionList(LauncherConstants.PROPERTIES.getVersionManifest(), proxy)), Agent.MINECRAFT, MinecraftReleaseTypeFactory.instance(), 21);
+        try {
+            Class<?> authClass = cl.loadClass("me.nepnep.msa4legacy.patches.MicrosoftAuthImpl");
+            this.msAuth = (MicrosoftAuth) authClass.newInstance();
+        } catch (InstantiationException e) {
+            throw new RuntimeException(e);
+        } catch (IllegalAccessException e) {
+            throw new RuntimeException(e);
+        } catch (ClassNotFoundException e) {
+            throw new RuntimeException(e);
+        }
+        this.launchDispatcher = new GameLaunchDispatcher(this, this.processArgs(args), this.msAuth);
         this.profileManager = new ProfileManager(this);
         ((SwingUserInterface)this.userInterface).initializeFrame();
         this.refreshVersionsAndProfiles();
@@ -211,7 +222,7 @@
     }
 
     private MinecraftUserInterface selectUserInterface(JFrame frame) {
-        return new SwingUserInterface(this, frame);
+        return new SwingUserInterface(this, frame, msAuth);
     }
 
     public com.mojang.launcher.Launcher getLauncher() {
@@ -227,9 +238,15 @@
     }
 
     public void ensureLoggedIn() {
-        UserAuthentication auth = this.profileManager.getAuthDatabase().getByUUID(this.profileManager.getSelectedUser());
+        this.msAuth.addAllToDatabase(null);
+        AuthenticationDatabase authDb = profileManager.getAuthDatabase();
+        String selected = profileManager.getSelectedUser();
+        UserAuthentication auth = authDb.getByUUID(selected);
+        MicrosoftAccount msa = authDb.getMSAByUUID(selected);
         if (auth == null) {
-            this.getUserInterface().showLoginPrompt();
+            if (msa == null) {
+                this.getUserInterface().showLoginPrompt();
+            }
         } else if (!auth.isLoggedIn()) {
             if (auth.canLogIn()) {
                 try {
@@ -240,7 +257,6 @@
                     catch (IOException e) {
                         LOGGER.error("Couldn't save profiles after refreshing auth!", (Throwable)e);
                     }
-                    this.profileManager.fireRefreshEvent();
                 }
                 catch (AuthenticationException e) {
                     LOGGER.error("Exception whilst logging into profile", (Throwable)e);
@@ -259,7 +275,6 @@
                 catch (IOException e) {
                     LOGGER.error("Couldn't save profiles after refreshing auth!", (Throwable)e);
                 }
-                this.profileManager.fireRefreshEvent();
             }
             catch (InvalidCredentialsException e) {
                 LOGGER.error("Exception whilst logging into profile", (Throwable)e);
@@ -269,6 +284,7 @@
                 LOGGER.error("Exception whilst logging into profile", (Throwable)e);
             }
         }
+        this.profileManager.fireRefreshEvent();
     }
 
     public UUID getClientToken() {
@@ -297,10 +313,12 @@
                 referencedObjects.add(object.getHash().toLowerCase());
             }
         }
-        File[] directories = objectsDir.listFiles(DirectoryFileFilter.DIRECTORY);
+        // Weird patch to fix recompilation
+        File[] directories = objectsDir.listFiles((FileFilter) DirectoryFileFilter.DIRECTORY);
         if (directories != null) {
             for (File directory : directories) {
-                File[] files = directory.listFiles(FileFileFilter.FILE);
+                // Weird patch to fix recompilation
+                File[] files = directory.listFiles((FileFilter) FileFileFilter.FILE);
                 if (files == null) continue;
                 for (File file : files) {
                     if (referencedObjects.contains(file.getName().toLowerCase())) continue;
@@ -373,7 +391,8 @@
         if (!virtualsDir.isDirectory()) {
             return;
         }
-        File[] directories = virtualsDir.listFiles(DirectoryFileFilter.DIRECTORY);
+        // Weird patch to fix recompilation
+        File[] directories = virtualsDir.listFiles((FileFilter) DirectoryFileFilter.DIRECTORY);
         if (directories != null) {
             for (File directory : directories) {
                 File lastUsedFile = new File(directory, ".lastused");
@@ -398,10 +417,12 @@
         if (!root.isDirectory()) {
             return;
         }
-        File[] versions = root.listFiles(DirectoryFileFilter.DIRECTORY);
+        // Weird patch to fix recompilation
+        File[] versions = root.listFiles((FileFilter) DirectoryFileFilter.DIRECTORY);
         if (versions != null) {
             for (File version : versions) {
-                File[] files = version.listFiles(FileFilterUtils.and(new PrefixFileFilter(version.getName() + "-natives-"), ageFilter));
+                // Weird patch to fix recompilation
+                File[] files = version.listFiles((FileFilter) FileFilterUtils.and(new PrefixFileFilter(version.getName() + "-natives-"), ageFilter));
                 if (files == null) continue;
                 for (File folder : files) {
                     LOGGER.debug("Deleting " + folder);
@@ -488,7 +509,6 @@
                 if (!FileUtils.deleteQuietly(file)) break block0;
                 LOGGER.info("Deleted empty directory {}", file);
             }
-            break;
         }
     }
 
@@ -521,4 +541,3 @@
         LOGGER = LogManager.getLogger();
     }
 }
-
